Feature: Add task detail pages and show objects yielded by normalizer

TwistMeister opened the pull request:
### Changes
Feature: Add task detail pages and show objects yielded by normalizer.

### Issue link
_Please add a link to the issue after "Closes". If there is no issue for this PR, please add it to the project board directly._

Closes #398 and #1288 

### Proof
<img width="1080" alt="image" src="https://github.com/minvws/nl-kat-coordination/assets/16188579/f5d5d44d-b34c-4268-871d-8c9374601a57">

---

### Code Checklist
- [ ] All the commits in this PR are properly PGP-signed and verified;
- [ ] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [ ] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [ ] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

underdarknl commented on octopoes/octopoes/repositories/origin_repository.py:
```diff
@@ -114,9 +114,6 @@ def save(self, origin: Origin, valid_time: datetime) -> None:
-        if old_origin == origin:
-            return
-
         self.session.add((XTDBOperationType.PUT, self.serialize(origin), valid_time))
```
Since we are not adding a new origin to the database, Should we not also change the previous origin (or old_origin) to be deleted from this moment in time?

underdarknl replied:
Currently an OOI is discovered by a normalizer and added to the database. At that point the origin record is also created, and points to the ooi and the normalizer job that yielded that object.
On subsequent runs on the same boefje/normalizer combination no update would be done to the database, the object would simply keep existing as no valid_till was present. 

To be able to show what that second run 'proofed' we need to make sure we either store the output of each normalizer run somewhere (bytes perhaps), or update the origin record to now point at this most recent proof.
By doing the latter we end up with one of two  things depending on the specific implementation:
* An origin record that when seen trough time/history contains a list of all jobs ever involved in proofing this ooi's existence
* A set of origin records related to this OOI, which all exists end to end, and collectively are the historical record of which jobs are involved in proofing this OOI.
I don't know how origin records are currently implemented, so which one it is needs to be checked, and implemented for accordingly

dekkers replied:
We have some documentation at https://docs.openkat.nl/developer_documentation/octopoes.html#origin
The current Origin model:
https://github.com/minvws/nl-kat-coordination/blob/c51122df60d4556214f898fbdb12072f24fdb6a6/octopoes/octopoes/models/origin.py#L9-L40

Looking at the documentation and the code:
- The Origin model  is used for declarations (objects created through rocky), observations (normalizer output) and inferences (bit output)
- The Origin model has an optional task_id that is only used by normalizers. The `__eq__` of the Origin method currently ignores task_id which causes it to not add a new Origin object if the other fields are the same. The purpose of the Origin objects is to keep track of the origins of objects in the graph and for this we don't need to track every normalizer task, so that hasn't been a problem up to now.
- This proposed change will cause a lot of duplicate inference origins for every bit run that has the same result, so we should look at another way to achiee this.

If we want to track all normalizer tasks in the origins/graph I think we have two options to do that:

1. We change the `__eq__` to include the task_id which will do what the current change intends to do. This will create a new Origin for every normalizer run, but won't cause duplication of inference Origins. It will create a lot more Origins object than we have now and also duplicate observation Origins in the sense that they have the same source/result. I'm not sure whether there isn't code that assumes that that doesn't happen which we would have to fix. Given that it wasn't possible before we also don't have unit tests that would test this, so we should investigate that if implement it this way. There might also be performance issues when we have a lot more Origin objects. 
2. Change the Origin object to include a list of tasks instead of a single task. If the result is the same we just add the task to the lists of tasks of the already existing Origin object. The challenge here is how to change this without breaking existing installations. Given that the task_id currently just points to the first normalizer task that returned this result, I am not sure whether we are really using this information anywhere, so we might be able to just change this without a problems.

Personally I have a preference for option 2, because it fits better with how we use the Origin object and we don't risk performance issues by creatings lots of Origin objects in the database.

dekkers replied:
This has been implemented

underdarknl commented on rocky/rocky/views/bytes_raw.py:
```diff
@@ -37,6 +37,15 @@ def get(self, request, **kwargs):
             return redirect(request.META["HTTP_REFERER"])
 
 
+# class NormalizerOriginRawView(OrganizationView):
```
Stale edit?

TwistMeister replied:
Yes, will remove

zcrt commented:
Nice! Would it be possible to (also) add (part of) this detailed information by expanding rows of the task table? (Just as currently with rows of the findings table.)

ammar92 commented on rocky/rocky/views/task_detail.py:
```diff
@@ -0,0 +1,79 @@
+            {
+                "url": reverse(
+                    "boefje_task_view",
+                    kwargs={"organization_code": self.organization.code, "task_id": uuid.UUID(context["task"].id).hex},
```
The key `task_id` already exists in the super's context, so you can use that instead of recreating a UUID object

TwistMeister replied:
Obviously. Thanks for pointing this out.

ammar92 commented on rocky/rocky/views/task_detail.py:
```diff
@@ -0,0 +1,79 @@
+    def get_output_oois(self, task):
+        try:
+            return self.octopoes_api_connector.list_origins(
+                valid_time=task.p_item.data.raw_data.boefje_meta.ended_at, task_id=uuid.UUID(task.id).hex
```
See my previous point about reusing vars

TwistMeister replied:
How can I reuse the context variable here? Or should I store the value in another way so I'm able to access it here as well?

ammar92 commented on rocky/rocky/views/task_detail.py:
```diff
@@ -0,0 +1,79 @@
+            {
+                "url": reverse(
+                    "normalizer_task_view",
+                    kwargs={"organization_code": self.organization.code, "task_id": uuid.UUID(context["task"].id).hex},
```
See my previous points

Darwinkel commented:
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

First of all, awesome feature!

### What works:
* The new pages
* No apparent regressions

### What doesn't work:
- `Download meta and raw data` button on the detail page for normalizers results in a 500:
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/7673348c-1aad-4acb-a87a-6f1a1b7bc691)
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/c68900fb-4e42-4f62-b57a-5c2068aa5ba7)

- Links can refer to non-existent boefjes
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/a4152552-e3be-469e-8487-35d54a42ba18)

- The column called "RAW Produced by Boefje" actually contains links to the boefjes' katalogus page, and not the raw output. 
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/4d311b9f-3082-4848-9c87-6a95a4d99bb2)


### Bug or feature?:

- From a design perspective, there are now two links to  `Download meta and raw data`, one on the tasks list, and one on the detail page. Isn't this a bit redundant / clutter?

![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/5df17935-af04-48b1-854b-824c05a0778e)
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/b3eeff74-1083-4603-bf33-ad2276e5f66e)

In addition, normalizers also have two links to download the raw input. Same point applies here.

TwistMeister commented:
@Darwinkel the link to the 'manual' boefje was broken already, so no regression there. I will create a fix for it on another branch, pointing at main.

The column "RAW created by boefje" isn't meant to include the RAW file, it's meant to state which boefje yielded said RAW file. I will put the column headers up for discussion with the UX-team so this will be more clear.

Having two download links, both in the list as on the detail page doesn't seem to troubling to me from a UX perspective. For design purposes it would be neat if we could remove it as a column from the list page, that I agree on. I will put this up for discussion as wel.

TwistMeister commented:
@zcrt While I think that's certainly possible and I do think something like that will be beneficial I would like to merge this as is, and not make the PR any bigger. I will take up the whole (re)design of the task lists and the detail pages with the UX-team as a whole instead of fixing this now and having to fix it again after the redesign.

Darwinkel commented:
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

Nice UX, I like the expandable menu.  

### What works:
- Rescheduling
- Clicking yielded objects
- Downloading completed raw and meta

### What doesn't work:
- Downloading dispatched meta
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/e4a0474f-e200-4574-b382-ad85b4662883)

- Rescheduling dispatched boefjes
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/da4ffba6-d24a-4db3-b4d6-e773ce7b7c51)

- Rescheduling dispatched normalizers
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/53020b50-04b4-4a59-a808-a1ae48cda848)

- When on a second page, a JavaScript popup occurs when clicking the expand button on the normalizers page
![image](https://github.com/minvws/nl-kat-coordination/assets/9536436/c027ba32-da5a-4253-a8f8-00df1b7b5255)

- When on a second page, I get a JSON response when clicking a normalizer task

underdarknl commented on rocky/assets/js/renderNormalizerOutputOOIs.js:
```diff
@@ -0,0 +1,45 @@
+
+  button.addEventListener("click", () => getJson(json_url, function(err, data) {
+    if (err !== null) {
+      alert('Somthing went wrong: ' + err);
```
This could just be moved to the getJson function, no need to pass along the err variable, and then act on it here.
Also, maybe do a console.log of the error, and some visual feedback for the user telling him the 'fold open' did not go as planned?

TwistMeister replied:
Yes, this definitely needs another fallback then to just show an alert.

underdarknl commented on rocky/assets/js/renderNormalizerOutputOOIs.js:
```diff
@@ -0,0 +1,45 @@
+buttons.forEach((button) => {
+  const raw_task_id = button.closest('tr').getAttribute('data-task-id');
+  const task_id = button.closest('tr').getAttribute('data-task-id').replace(/-/g, "");
+  const json_url = window.location.href + "/" + task_id;
```
If this JS is going to be used on other pages too, this maybe just needs to be a constant?

TwistMeister replied:
Not sure what you mean. The task_id is only known within the context of the button, right?

underdarknl commented on rocky/assets/js/renderNormalizerOutputOOIs.js:
```diff
@@ -0,0 +1,45 @@
+          const url = window.location.href.replace('/tasks/normalizers', '')
+          element.innerHTML = "<a href='" + url + "/objects/detail/?ooi_id=" + object +"'>" + object + "</a>"
+
+          button.closest('tr').nextElementSibling.querySelector('#yielded-objects-'+raw_task_id).appendChild(element);
```
What happens if I open, and then close, and then open the same row again? Im guessing we'd end up with the same yielded objects listed twice?

TwistMeister replied:
Thanks for pointing this out. I had already fixed this, but had not yet pushed my changes :)

TwistMeister commented:
@Darwinkel I have addressed the points you raised. The points I didn't fix in this PR, since it wasn't regression are:
- Downloading dispatched meta results in an error. Follow-up ticket: https://github.com/minvws/nl-kat-coordination/issues/1523
- Rescheduling task "twice" results in an error. Follow-up ticket: https://github.com/minvws/nl-kat-coordination/issues/1589

What I did fix is that a task that isn't yet completed doesn't show the "reschedule" button. But if you click "reschedule" on an earlier instance of that task, while a new one is already running the error you observed persists, hence ticket https://github.com/minvws/nl-kat-coordination/issues/1589.

Darwinkel commented:
Issue has been fixed! 

Unrelated: I took the liberty to add `--use-fuzzy` to some of our `compilemessages` pipelines as I discovered that we have been forgetting/losing translations for a while now. Django sometimes tries to automatically infer/fix new translation strings, which cause them to be annotated as fuzzy. By default, `compilemessages` doesn't include these, which has the subtle effect of effectively discarding some of our previous translation work.
